process/smartContract/scQueryService.go

```diff
@@ -168,6 +241,124 @@ func (service *SCQueryService) executeScCall(query *process.SCQuery, gasPrice ui
 	return vmOutput, nil
 }
 
+func (service *SCQueryService) extractBlockRootHash(query *process.SCQuery) ([]byte, error) {
+	if len(query.BlockRootHash) > 0 {
+		return query.BlockRootHash, nil
+	}
+
+	if len(query.BlockHash) > 0 {
+		blockHeader, err := service.getBlockHeaderByHash(query.BlockHash)
+		if err != nil {
+			return make([]byte, 0), err
+		}
+
+		return service.getBlockRootHash(query.BlockHash, blockHeader), nil
+	}
+
+	if query.BlockNonce.HasValue {
+		blockHeader, blockHash, err := service.getBlockHeaderByNonce(query.BlockNonce.Value)
+		if err != nil {
+			return make([]byte, 0), err
+		}
+
+		return service.getBlockRootHash(blockHash, blockHeader), nil
+	}
+
+	return make([]byte, 0), nil
+}
+
+func (service *SCQueryService) getBlockHeaderByHash(headerHash []byte) (data.HeaderHandler, error) {
+	epoch, err := service.getOptionalEpochByHash(headerHash)
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := service.getBlockHeaderInEpochByHash(headerHash, epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	return header, nil
+}
+
+func (service *SCQueryService) getOptionalEpochByHash(hash []byte) (core.OptionalUint32, error) {
+	if !service.historyRepository.IsEnabled() {
+		return core.OptionalUint32{}, nil
+	}
+
+	epoch, err := service.historyRepository.GetEpochByHash(hash)
+	if err != nil {
+		return core.OptionalUint32{}, err
+	}
+
+	return core.OptionalUint32{Value: epoch, HasValue: true}, nil
+}
+
+func (service *SCQueryService) getBlockHeaderInEpochByHash(headerHash []byte, epoch core.OptionalUint32) (data.HeaderHandler, error) {
+	shardId := service.shardCoordinator.SelfId()
+	unitType := dataRetriever.GetHeadersDataUnit(shardId)
+	storer, err := service.storageService.GetStorer(unitType)
+	if err != nil {
+		return nil, err
+	}
+
+	var headerBuffer []byte
+
+	if epoch.HasValue {
+		headerBuffer, err = storer.GetFromEpoch(headerHash, epoch.Value)
+	} else {
+		headerBuffer, err = storer.Get(headerHash)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := process.UnmarshalHeader(shardId, service.marshaller, headerBuffer)
+	if err != nil {
+		return nil, err
+	}
+
+	return header, nil
+}
+
+func (service *SCQueryService) getBlockRootHash(headerHash []byte, header data.HeaderHandler) []byte {
+	blockRootHash, err := service.scheduledTxsExecutionHandler.GetScheduledRootHashForHeaderWithEpoch(
```

Comment:
OK I see why you added the scheduled execution handler, but you can use instead getBlockHeaderByNonce and request the header.Nonce+1 and check on that the scheduledRootHash.

Reply:
We can discuss as well, I see that this is similar to how it is done for on nodeBlocks.go
This code appears duplicated, maybe it would be better to extract a component and reuse it in both places instead.

Reply:
added a TODO for the duplicated code